### ISSUE TITLE
Add support for string translation for names, descriptions, and messages

### ISSUE THREAD

**Issue by TravisCarden** [php-tuf/composer-stager#123]
As surfaced in https://github.com/php-tuf/composer-stager/pull/118, Drupal requires strings that will appear in user interfaces to be translatable for multilingual support. But it depends on them being defined using its [Localization API](https://www.drupal.org/node/322729), which obviously, Composer Stager knows nothing about. (Nor would it matter, as far as I know, since the code still wouldn't be in the Drupal codebase for discovery.) We need to solve this problem for dependent projects in a way that accounts for the complexity of substitutions, for example.

### Background

> @larowlan, https://github.com/php-tuf/composer-stager/pull/118 - How would downstream users of this go about translating the precondition names/descriptions/unfulfilled/fulfilled strings. For use with e.g. Drupal this would be a requirement I'd think

> @TravisCarden https://github.com/php-tuf/composer-stager/pull/118#issuecomment-1482125783 - This is the hard one, isn't it? I'm glad you bring it up. We discussed it early on, but there wasn't consensus on whether it was actually an issue. I seems to me that it would be a requirement. But as you imply, there are obvious difficulties, and I don't have an easy answer. A few possibilities come readily to my mind:
>
> 1. Composer Stager could provide very specific exception codes (c.f. https://github.com/php-tuf/composer-stager/issues/71), and Drupal could declare its own translatable strings based on them instead of just taking whatever it gets back. The major drawback to this, of course, is that details will be lost, such as error messages that state what went wrong and specify a file it occurred in.
>
> 1. We could add the ability to pass some sort of translation callback into Composer Stager API calls.
>
> 1. Instead of returning strings, we could have Composer Stager return some sort of object that provides placeholder substitution similar to Drupal's `TranslatableMarkup`.

**Related issue:** [Support scanning Composer Stager library for translatable strings [#3354011]](https://www.drupal.org/project/potx/issues/3354011) with the [Translation template extractor (`potx`)](https://www.drupal.org/project/potx).

**Note:** Adding this to https://github.com/php-tuf/composer-stager/milestone/1 as a requirement for a stable release.

**Comment by larowlan** [comment]
Made a possible start in https://github.com/php-tuf/composer-stager/pull/124 - if that looks reasonable can continue fixing unit tests

**Comment by goba** [comment]
I don't have a strong opinion either way on the API (which is why I am not a framework manager in Drupal :D). I think either passing a translation service of sorts to composer stager to generate the messages with could work (where composer stager would be responsible for using that service API properly, and Drupal could inject its translations ervice). Or where composer stager returns some structured metadata about messages it needs to expose including well defined constant values for the particular message and metadata about replacement values (file names, etc), in which case Drupal would be responsible to have a list of messages and placeholder handling for the messages.

A clear benefit of the later approach would be that the Drupal codebase would have a list of translatable strings for localize.drupal.org, etc. I am not sure how to achieve that with the former approach unless composer stager can use some standar markers for such strings or explicitly have a list of translatable strings maintained.

**Comment by larowlan** [comment]
So would it be better to just have an InterfaceStringsInterface with constants for all the messages? And l.d.o could use that for finding strings to translate?

**Comment by larowlan** [comment]
So in theory we could write a potx patch to work with a list of constants on a third-party interface?

**Comment by TravisCarden** [comment]
So, I initially put forward the possibility of using constants; but as I research the matter, I wonder if this needs to be even that complicated. It seems to me we could just change any string "getters" in Composer Stager to return `\Symfony\Contracts\Translation\TranslatableInterface` objects. (Drupal already includes the interface via [Symfony Translation Contracts](https://github.com/symfony/translation-contracts).) We could instantiate them with a factory service that Drupal could override to create `TranslatableMarkup` objects instead, simultaneously providing a method for the Translation template extractor (`potx`) to search for for https://localize.drupal.org/--which would just need to include the Composer Stager directory in `vendor` in its scanning. All things considered, I think this approach has the best balance of strengths and weaknesses and would have fairly minimal API impact. Unless someone has a better proposal or a reason against this one, I would like to start drilling into this approach. Thoughts?

**Comment by larowlan** [comment]
It sounds like its worth exploring

**Comment by TravisCarden** [comment]
I created [Support scanning Composer Stager library for translatable strings [#3354011]](https://www.drupal.org/project/potx/issues/3354011) with the Translation template extractor (`potx`).

**Comment by mxr576** [comment]
Subscribe :)

Exception translation is always an interesting topic, also translating strings from 3rd party (plain PHP) libraries.

> Drupal could then just call a new TranslatableMarkup($e->getMessage()) known that all the values for the message have already been collected.

Personally, I do not like this idea. Exceptions and their content may not be meant to be exposed to end users. They can contain internal details, even information that a simple user MUST NOT see. 

This seemed a better approach to me if that design idea wins:

> Composer Stager could provide very specific exception codes (c.f. https://github.com/php-tuf/composer-stager/issues/71), and Drupal could declare its own translatable strings based on them i

Regarding...

> It seems to me we could just change any string "getters" in Composer Stager to return \Symfony\Contracts\Translation\TranslatableInterface objects. 

I am not sure what "getters" refers to in this sentence exactly. My 2 cents is that if the hexagonal arch (ports and adapters) principle was applied to this package if I would be the maintainer, I would even try to avoid depending on Symfony contracts if it does not look seems to be an overkill. I would define a port (contract/interface) for translation support and 3rd party dependencies can define the adapter for that... but I am still unsure is a plain PHP library should care about multilingual support at this level. So maybe just throwing _meaningful_ exceptions is the best idea and the code that catches them can figure out how to turn them into human-readable error messages.

**Comment by TravisCarden** [comment]
@mxr576 I had hoped to be actually responding to your comments right now, but as I've been considering them I realize I need to do a little more investigation. I hope to reply very soon. I just wanted you to know I'm not ignoring you.

**Comment by TravisCarden** [comment]
Okay, @mxr576, that didn't take as long as I thought it would.

tl;dr: I think we do need to take responsibility for string translation, but I agree with you on the hexagonal design aspect, i.e., defining our own interfaces. Here are the details:

> I am still unsure is a plain PHP library should care about multilingual support at this level.

I think it has been established that this is a requirement for this reason: the question is not merely, "Should Composer Stager take responsibility for translation or should we leave that up to the projects that use it?" It is, "Is it even _possible_ for other projects to do it if Composer Stager doesn't?" And I think the conclusion we've arrived at is that if Composer Stager doesn't do it, it actually _prevents_ other projects from doing it for themselves. The main reason for that is variable substitution. Some strings include variable values, like the path to a file that's preventing a domain operation. If Composer Stager only returns simple strings, either they'll be essentially untranslatable or they'll lose that "metadata": If strings are returned with variable substitution already done, they can't be translated at all (without some kind of unholy regex-based extraction). If clients are left to depend on exception codes, they won't have access to that variable data; and help messages, for example, will lose the ability to tell site owners _what_ they need to fix, just that they need to fix _something_.

>> Drupal could then just call a new TranslatableMarkup($e->getMessage()) known that all the values for the message have already been collected.
>
> Personally, I do not like this idea. Exceptions and their content may not be meant to be exposed to end users. They can contain internal details, even information that a simple user MUST NOT see.

You're reacting to the idea of translating _exceptions_ specifically here, because they should never be shown to users, right? That would be a matter of how dependent projects _use_ translatable strings returned by Composer Stager, not of _what_ Composer Stager returns, so it shouldn't actually influence the design decisions here one way or the other.

> I am not sure what "getters" refers to in this sentence exactly.

I was referring to getters for UI strings, like the name, description, and status message of preconditions, i.e., `Precondition::getName()`, `Precondition::getDescription()`, and `Precondition::getStatusMessage()`. Since those are explicitly _meant_ for user display, it seems obvious to me that they would return translatable objects. In particular, `::getStatusMessage()` requires variable substitution, which is why this isn't just about exceptions.

> I would even try to avoid depending on Symfony contracts if it does not look seems to be an overkill. I would define a port (contract/interface) for translation support and 3rd party dependencies can define the adapter for that...

This is the point I've been wrestling with, and I've actually reversed my original thinking rubber ducking it just now. I think we should define our own interfaces (ports) as you say, but make them identical to the (widely-used, battle-tested) `symfony/translation-contracts` ones. Dependent projects will be able to use our interfaces with _paper thin_ adapters, if any. Having dug around a bit, I think Drupal would require literally nothing but an addition to a `services.yml`, if that, and the modification to the Drupal Translation template extractor mentioned before (https://www.drupal.org/project/potx/issues/3354011). We should still require the `symfony/translation-contracts` package, but only to use its `TranslatorTrait` internally (i.e., in the infrastructure layer) so we don't have to reinvent the wheel. As stated before, this won't result in adding a dependency to Drupal, which already requires it indirectly.

I think this satisfies your concerns, @mxr576, and I'm very happy with it. Thanks for the helpful feedback. I'm going to proceed accordingly in my in-progress PR (#141), which will require just a little refactoring for my updated thinking but is a good way along overall.

**Comment by effulgentsia** [comment]
I don't think any of the names/descriptions/messages returned by Composer Stager contain HTML, so I don't think Composer Stager has  a need for its interface to support Stringable as a return. So the question is: is it a good idea to do even without a use-case?

**Comment by catch56** [comment]
I don't think it's about HTML, it's about the lazy rendering of the string including replacing placeholders via TranslatableMarkup

**Comment by TravisCarden** [comment]
With some review from @effulgentsia and at his suggestion, I've gone ahead and merged the PR. Additional feedback can be provided in follow-ups or during the overall review of the codebase before including it in Drupal.
